File: uf-client-service/uf-client-service-api/src/main/java/com/kynetics/uf/android/api/UFServiceConfiguration.java
/*
 * Copyright © 2017-2018  Kynetics  LLC
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  which accompanies this distribution, and is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 */

package com.kynetics.uf.android.api;

import java.io.Serializable;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

/**
 * @author Daniele Sergio
 */
public class UFServiceConfiguration implements Serializable{

    public static class Builder {
        private Builder(){
            targetAttributes = new HashMap<>(2);
            targetAttributes.put("client","Android");
            final SimpleDateFormat sm = new SimpleDateFormat("yyyy.MM.dd HH:mm:ss z", Locale.getDefault());
            targetAttributes.put("date", sm.format(new Date()));

        }
        public Builder withTenant(String tenant) {
            this.tenant = tenant;
            return this;
        }

        public Builder withControllerId(String controllerId) {
            this.controllerId = controllerId;
            return this;
        }

        public Builder withGetawayToken(String gatewayToken) {
            this.gatewayToken = gatewayToken;
            return this;
        }

        public Builder withTargetToken(String targetToken) {
            this.targetToken = targetToken;
            return this;
        }

        public Builder withRetryDelay(long retryDelay) {
            if(retryDelay >= 30_000) {
                this.retryDelay = retryDelay;
            }
            return this;
        }

        public Builder withUrl(String url) {
            this.url = url;
            return this;
        }

        public Builder withApiMode(boolean apiMode) {
            this.apiMode = apiMode;
            return this;
        }

        /**
         * @deprecated  As of release 0.3.4, replaced by {@link #withEnable(boolean)}
         */
        @Deprecated
        public Builder witEnable(boolean enable) {
            return withEnable(enable);
        }

        /**
         * @deprecated  As of release 0.3.4, replaced by {@link #withTargetAttributes(Map<String,String>)}
         */
        @Deprecated
        public Builder witArgs(Map<String,String> args) {
            return withTargetAttributes(args);
        }

        public Builder withEnable(boolean enable) {
            this.enable = enable;
            return this;
        }

        public Builder withTargetAttributes(Map<String,String> targetAttribute) {
            if(targetAttribute != null && targetAttribute.size() > 0) {
                this.targetAttributes = targetAttribute;
            }
            return this;
        }

        public Builder withIsUpdateFactoryServer(boolean isUpdateFactoryServer){
            this.isUpdateFactoryServer = isUpdateFactoryServer;
            return this;
        }

        public UFServiceConfiguration build() {
            Objects.requireNonNull(tenant);
            Objects.requireNonNull(controllerId);
            Objects.requireNonNull(url);
            if(retryDelay < 0 ){
                throw new IllegalStateException("retryDelay must be grater than 0");
            }
            return new UFServiceConfiguration(tenant, controllerId, retryDelay, url,
                    targetToken == null ? "" : targetToken,
                    gatewayToken == null ? "" : gatewayToken,
                    apiMode, enable, isUpdateFactoryServer,
                    targetAttributes);
        }

        public boolean configurationIsValid(){
            return notEmptyString(tenant)
                    && notEmptyString(controllerId)
                    && notEmptyString(url)
                    && retryDelay > 0;
        }

        private boolean notEmptyString(String stringToTest){
            return stringToTest != null && !stringToTest.isEmpty();
        }
        private String tenant;
        private String controllerId;
        private long retryDelay = 900_000;
        private String url;
        private boolean apiMode = true;
        private boolean enable = true;
        private boolean isUpdateFactoryServer = true;
        private String targetToken;
        private String gatewayToken;
        private Map<String,String> targetAttributes;
    }

    public static Builder builder(){
        return new Builder();
    }

    public String getTenant() {
        return tenant;
    }

    public String getControllerId() {
        return controllerId;
    }

    public long getRetryDelay() {
        return retryDelay;
    }

    public String getUrl() {
        return url;
    }

    public Boolean isApiMode() {
        return apiMode;
    }

    public Boolean isEnable() {
        return enable;
    }

    public String getTargetToken() {
        return targetToken;
    }

    public String getGatewayToken() {
        return gatewayToken;
    }

    /**
     * @deprecated  As of release 0.3.4, replaced by {@link #getTargetAttributes()}
     */
    @Deprecated
    public Map<String, String> getArgs() {
        return getTargetAttributes();
    }

    public Map<String, String> getTargetAttributes() {
        return targetAttributes;
    }

    public boolean isUpdateFactoryServe() {
        return isUpdateFactoryServe;
    }

    private UFServiceConfiguration(String tenant,
                                   String controllerId,
                                   long retryDelay,
                                   String url,
                                   String targetToken,
                                   String gatewayToken,
                                   boolean apiMode,
                                   boolean isEnable,
                                   boolean isUpdateFactoryServe,
                                   Map<String,String> targetAttributes) {
        this.tenant = tenant;
        this.controllerId = controllerId;
        this.retryDelay = retryDelay;
        this.url = url;
        this.targetToken = targetToken;
        this.gatewayToken = gatewayToken;
        this.apiMode = apiMode;
        this.enable = isEnable;
        this.isUpdateFactoryServe = isUpdateFactoryServe;
        this.targetAttributes = targetAttributes;
    }

    private final String tenant;
    private final String controllerId;
    private final long retryDelay;
    private final String url;
    private final String targetToken;
    private final String gatewayToken;
    private final boolean apiMode;
    private final boolean enable;
    private final Map<String,String> targetAttributes;
    private final boolean isUpdateFactoryServe;

    private static final long serialVersionUID = -6025361892414738765L;
}

File: uf-client-service/uf-client-service-api/src/main/kotlin/com/kynetics/uf/android/api/UFServiceConfiguration.kt
/*
 * Copyright © 2017-2020  Kynetics  LLC
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.kynetics.uf.android.api

import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration
import kotlinx.serialization.json.JsonException

@Serializable
/**
 * This class represent the [com.kynetics.uf.android.UpdateFactoryService]'s configuration
 *
 * @property tenant the tenant
 * @property controllerId id of the controller
 * @property retryDelay time to wait between poll after a connection error
 * @property url of update server
 * @property targetToken target token
 * @property gatewayToken gateway token
 * @property isApiMode true to ask user authorization sending message to client, false to use dialog
 * @property isEnable true to enable the service, false to disable it
 * @property isUpdateFactoryServe true when the server is an UpdateServer, false when the server is an
 *  Hawkbit server
 * @property targetAttributes target's tags
 */
data class UFServiceConfiguration(
    val tenant: String,
    val controllerId: String,
    @Deprecated("As of release 1.0.0-RC repalced by exponential backoff")
    val retryDelay: Long,
    val url: String,
    val targetToken: String,
    val gatewayToken: String,
    val isApiMode: Boolean,
    val isEnable: Boolean,
    val isUpdateFactoryServe: Boolean,
    val targetAttributes: Map<String, String>
) : java.io.Serializable {

    /**
     * Json serialization
     */
    fun toJson(): String {
        return json.stringify(serializer(), this)
    }

    /**
     * [UFServiceConfiguration]'s builder class
     *
     * @property tenant the tenant
     * @property controllerId id of the controller
     * @property retryDelay time to wait between poll after a connection error, default value is 900_000
     * @property url of update server
     * @property targetToken target token
     * @property gatewayToken gateway token
     * @property isApiMode true to ask user authorization sending message to client, false to use dialog, default value is true
     * @property isEnable true to enable the service, false to disable it, default value is true
     * @property isUpdateFactoryServe true when the server is an UpdateServer, false when the server is an default value is true
     *  Hawkbit server
     * @property targetAttributes target's tags
     */
    class Builder internal constructor() {

        private var tenant: String? = ""
        private var controllerId: String? = ""
        private var retryDelay: Long = 900000
        private var url: String? = ""
        private var apiMode = true
        private var enable = true
        private var isUpdateFactoryServer = true
        private var targetToken: String? = ""
        private var gatewayToken: String? = ""
        private var targetAttributes: Map<String, String> = mutableMapOf()
        /**
         * Configure the tenant parameter
         */
        fun withTenant(tenant: String?): Builder {
            this.tenant = tenant
            return this
        }

        /**
         * Configure the controller id  parameter
         */
        fun withControllerId(controllerId: String?): Builder {
            this.controllerId = controllerId
            return this
        }

        /**
         * Configure the gateway token  parameter
         */
        fun withGetawayToken(gatewayToken: String?): Builder {
            this.gatewayToken = gatewayToken
            return this
        }

        /**
         * Configure the target token parameter
         */
        fun withTargetToken(targetToken: String?): Builder {
            this.targetToken = targetToken
            return this
        }

        /**
         * Configure the retryDelay parameter
         */
        @Deprecated("As of release 1.0.0-RC removed")
        fun withRetryDelay(retryDelay: Long): Builder {
            this.retryDelay = retryDelay
            return this
        }

        /**
         * Configure the url parameter
         */
        fun withUrl(url: String?): Builder {
            this.url = url
            return this
        }

        /**
         * Configure the api mode parameter
         */
        fun withApiMode(apiMode: Boolean): Builder {
            this.apiMode = apiMode
            return this
        }

        /**
         * Configure the enable parameter
         * @see withEnable
         */
        @Deprecated("As of release 0.3.4 replaced by withEnable(boolean)",
                replaceWith = ReplaceWith(""))
        fun witEnable(enable: Boolean): Builder {
            return withEnable(enable)
        }

        /**
         * Configure the target attributes parameter
         *
         * @see withTargetAttributes
         */
        @Deprecated("As of release 0.3.4 replaced by withTargetAttributes(Map<String,String>",
                replaceWith = ReplaceWith(""))
        fun witArgs(args: Map<String, String>): Builder {
            return withTargetAttributes(args)
        }

        /**
         * Configure the enable parameter
         */
        fun withEnable(enable: Boolean): Builder {
            this.enable = enable
            return this
        }

        /**
         * Configure the target attributes parameter
         */
        fun withTargetAttributes(targetAttribute: Map<String, String>?): Builder {
            if (targetAttribute != null && targetAttribute.size > 0) {
                this.targetAttributes = targetAttribute
            }
            return this
        }

        /**
         * Configure the UpdateFactoryServer parameter
         */
        fun withIsUpdateFactoryServer(isUpdateFactoryServer: Boolean): Builder {
            this.isUpdateFactoryServer = isUpdateFactoryServer
            return this
        }

        /**
         * Build an instance of UFServiceConfigure with the configured parameters
         *
         * @throws IllegalStateException when the retryDelay parameter is lower then 0
         *
         */
        fun build(): UFServiceConfiguration {
            if (retryDelay < 0) {
                throw IllegalStateException("retryDelay must be grater than 0")
            }
            return UFServiceConfiguration(tenant ?: "", controllerId ?: "", retryDelay, url ?: "",
                    targetToken ?: "",
                    gatewayToken ?: "",
                    apiMode, enable, isUpdateFactoryServer,
                    targetAttributes)
        }

        /**
         * Validate the [UFServiceConfiguration] that is built using the [build] method
         */
        fun configurationIsValid(): Boolean {
            return (notEmptyString(tenant) &&
                    notEmptyString(controllerId) &&
                    notEmptyString(url) &&
                    retryDelay > 0)
        }

        private fun notEmptyString(stringToTest: String?): Boolean {
            return stringToTest != null && !stringToTest.isEmpty()
        }
    }

    companion object {
        private val json = Json(JsonConfiguration.Stable.copy(strictMode = false))
        private const val serialVersionUID = -6025361892414738765L

        /**
         * Instantiate a builder
         */
        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }

        /**
         * Deserializes given json [data] into a corresponding object of type [UFServiceConfiguration].
         * @throws [JsonException] in case of malformed json
         * @throws [SerializationException] if given input can not be deserialized
         */
        @JvmStatic
        fun fromJson(data: String): UFServiceConfiguration {
            return json.parse(serializer(), data)
        }
    }

    /**
     *  get target's tags
     *
     *  @see targetAttributes
     */
    @Deprecated("As of release 0.3.4 replaced by targetAttributes",
            replaceWith = ReplaceWith(""))
    fun getArgs(): Map<String, String> {
        return this.targetAttributes
    }
}
